specs/core/1_shard-data-chains.md

```diff
@@ -122,3 +131,118 @@ Verify that the `shard_block_combined_data_root` is the output of these function
+* Verify that `hash(new_commitment) = state.validators[index].proof_of_custody_commitment`.
+* Verify that `block.slot >= state.validators[index].proof_of_custody_last_change + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`.
+* Set `state.validators[index].proof_of_custody_second_last_change` to `state.validators[index].proof_of_custody_last_change`.
+* Set `state.validators[index].proof_of_custody_last_change` to `block.slot`.
+
+#### PROOF_OF_CUSTODY_CHALLENGE
+
+```
+{
+    'attestation': SpecialAttestationData,
+    'validator_index': 'uint64',
+    'data_index': 'uint64',
+}
+```
+
+Perform the following checks:
+
+* Verify that the attestation is valid.
+* Verify that `attestation.slot > state.validators[index].proof_of_custody_second_last_change`.
+* Verify that `validator_index in attestation.aggregate_sig_poc_0_indices` or `validator_index in attestation.aggregate_sig_poc_1_indices`; if the former, let `bit = False`, else `bit = True`.
+* Verify that `state.validators[index].status == ACTIVE` or `state.validators[index].status == PENDING_EXIT` and `block.slot < state.validators[index].exit_slot + PROOF_OF_CUSTODY_MIN_CHANGE_PERIOD`
+* Let `check_value = as_uint256(hash(as_bytes8(validator_index) + as_bytes8(attestation.slot) + as_bytes8(attestation.shard_id)))`.
+* Let `FULL_PROOF_OF_CUSTODY_VALCOUNT = TARGET_COMMITEE_SIZE * SHARD_COUNT`, and `target = 2**256 * attestation.total_validator_count // FULL_PROOF_OF_CUSTODY_VALCOUNT`.
+* Verify that `check_value <= target`.
+
+* Let `seed_hash = state.validators[index].proof_of_custody_commitment if attestation.slot > state.validators[index].proof_of_custody_last_change else hash(state.validators[index].proof_of_custody_commitment)`.
+* Append to `state.proof_of_custody_challenges` the object `ProofOfCustodyChallenge(responder_index=validator_index, seed_hash=seed_hash, depth=attestation.proof_of_custody_depth, data_root=attestation.data.shard_block_hash, data_index=data_index, expiry_slot=block.slot+PROOF_OF_CUSTODY_RESPONSE_DEADLINE, challenger_index=get_proposer(state, block), bit=bit)`.
+
+A block can have maximum one proof of custody challenge, and it must appear before all `PROOF_OF_CUSTODY_SEED_CHANGE` objects.
+
+#### PROOF_OF_CUSTODY_RESPONSE
+```
+{
+    'challenge_index': 'uint64',
+    'leaf': 'hash32',
+    'root': 'hash32',
+    'seed': 'hash32',
+    'data_merkle_branch': '[hash32]',
+    'poc_merkle_branch': '[hash32]',
+}
+```
+Perform the following checks:
+
+* Verify that `challenge_index < len(state.proof_of_custody_challenges)`.
+* Let`challenge = state.proof_of_custody_challenges[challenge_index]`.
+* Verify that `hash(seed) = challenge.seed_hash`
+* Verify that `verify_merkle_branch(leaf, data_merkle_branch, challenge.depth, challenge.data_index, root)` passes.
+* Verify that `verify_merkle_branch(leaf, poc_merkle_branch, challenge.depth, challenge.data_index, xor(root, seed))` passes.
+* Verify that `uint256(root) % 2 == challenge.bit`.
```

Comment:
does this count as big int arithmetic? Can we do a `&` with a `MASK` to pull out the bit instead?